src/tiled/abstracttiletool.cpp

```diff
@@ -40,6 +40,7 @@ AbstractTileTool::AbstractTileTool(const QString &name,
                                    BrushItem *brushItem,
                                    QObject *parent)
     : AbstractTool(name, icon, shortcut, parent)
+    , mRandomCacheValid(false)
```

Comment:
I'm not sure about putting this bool here, since not all tile tools have a random cache (actually only the `AbstractTileFillTool` and `StampBrush` keep around their `RandomPicker` instance). I think I'd currently prefer only those classes to include the bool and live with a bit of code duplication for now. Note that `mapDocumentChanged` is virtual so they can override it for connecting to the `tileProbabilityChanged` signal.

Reply:
I'm moved it to `StampBrush`. After my implementation there is approved I can do something similar to the `AbstractTileFillTool`.